Use DynamicDependencyAttribute

sbomer opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/37475. I'm still waiting for [dependency flow](https://github.com/dotnet/runtime/pull/37747) (https://github.com/mono/linker/pull/1262 allows keeping members by name like `PreserveDependency` did), but opening this for review in the meantime.

sbomer committed:
Use DynamicDependencyAttribute

stephentoub commented on src/libraries/System.Net.HttpListener/src/System/Net/Windows/CookieExtensions.cs:
```diff
@@ -13,7 +13,7 @@ internal static class CookieExtensions
         private static Func<Cookie, string> s_toServerStringFunc;
 
-        [PreserveDependency("ToServerString", "System.Net.Cookie", "System.Net.Primitives")]
+        [DynamicDependency("ToServerString", "System.Net.Cookie", "System.Net.Primitives")]
```
Can this use typeof(Cookie)? Same below.

marek-safar replied:
The attribute is should not be needed at all

MichalStrehovsky commented:
Should we use this as an opportunity to clean these up?

* There's a lot of annotations that shouldn't be needed because the reflection pattern is within the method and the linker can analyze that.
* Some annotations should have a TODO mentioning https://github.com/mono/linker/issues/943
* Some could be converted to `DynamicallyAccessedMembers`, which is a preferable annotation because it works with the dataflow analysis.

marek-safar commented:
> There's a lot of annotations that shouldn't be needed because the reflection pattern is within the method and the linker can analyze that.

Yes, let's remove them

> DynamicallyAccessedMembers, which is a preferable annotation because it works with the dataflow analysis.

I disagree at BCL level we should be explicit where possible

MichalStrehovsky commented:
> > DynamicallyAccessedMembers, which is a preferable annotation because it works with the dataflow analysis.

> I disagree at BCL level we should be explicit where possible

I meant places like the `GetHelperType` method in this pull request: if you annotate the return parameter with `DynamicallyAccessedMembers`, it both ensures we keep the constructor, and the place that passes the result of `GetHelperType` to `Activator` stops generating warning. If we keep DynamicDependency, we need to add a suppression to the `Activator` call here:

https://github.com/dotnet/runtime/blob/b8339a8c61aa18cee59ea58cdf9eb1376b581343/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/CallInstruction.cs#L239

vitek-karas commented:
I would probably postpone converting `DynamicDependency` to data flow annotations into a different PR/issue. I think it's a good idea to do that where it makes sense, since it's the cleaner approach which works better as the code evolves.

For this conversion itself, there are basically 3 cases:
* `DynamicDependency` into a different assembly - we need to keep these, regardless of the reflection pattern. This is basically a workaround to the problem that linker can't add assemblies into the closure in `Type.GetType` and so on. So the attribute in this case mainly acts as the thing which pulls in the assembly.
* `DynamicDependency` to things within the same assembly which there are no reflection patterns to access - like the stuff on System.String for example. We also need to keep these, as there's no other way to handle this.
* `DynamicDependency` on things within the same assembly, where linker will figure out the pattern. We should remove these. Linker will issue a warning if it can't figure out the pattern. In such case adding `DynamicDependency` will actually not remove the warning, we would also have to add `UncoditionalSuppressMessage`.

eerhardt commented on eng/DefaultGenApiDocIds.txt:
```diff
@@ -27,7 +27,6 @@ T:System.Runtime.CompilerServices.IteratorStateMachineAttribute
 T:System.Runtime.CompilerServices.MethodImpl
 T:System.Runtime.CompilerServices.NullableAttribute
 T:System.Runtime.CompilerServices.NullableContextAttribute
-T:System.Runtime.CompilerServices.PreserveDependencyAttribute
```
Should we add `DynamicDependency` here? That way the `ref` assemblies won't start getting it when someone re-runs the generator.

eerhardt commented on src/libraries/System.Data.Common/src/System/Data/SQLTypes/SqlXml.cs:
```diff
@@ -128,7 +128,7 @@ internal static XmlReader CreateSqlXmlReader(Stream stream, bool closeInput = fa
         private static MethodInfo CreateSqlReaderMethodInfo
         {
-            [PreserveDependency("CreateSqlReader", "System.Xml.XmlReader", "System.Private.Xml")]
+            [DynamicDependency("CreateSqlReader", "System.Xml.XmlReader", "System.Private.Xml")]
```
Can this use `typeof(System.Xml.XmlReader)`?

eerhardt commented:
>> There's a lot of annotations that shouldn't be needed because the reflection pattern is within the method and the linker can analyze that.

> Yes, let's remove them

The plan was to be explicit with the attributes until we can turn on the linker analysis, and have a test that covers the dependency is kept. Once we have linker analysis enabled (and it breaking the build if this gets broken) and a test that covers these reflection usages, I'd prefer we keep being explicit.

sbomer committed:
PR feedback

- Use typeof in more places
- Add comments about DynamicallyAccessedMembers
- Add comments about cases where we work around linker limitations
- Also move AssemblyBuilder`s DynamicDependency closer to where it
is actually needed.

sbomer committed:
Add issue links

sbomer commented:
For now I have added comments about the ones that should be converted to `DynamicallyAccessedMembers` and those that are needed to work around https://github.com/mono/linker/issues/943. I'll leave removing the unnecessary ones for another PR when we turn on analysis warnings and maybe add tests. [This change](https://github.com/sbomer/runtime/compare/dynamicDependency...sbomer:removeDynamicDependency?expand=1) shows the ones we should be able to remove, and https://github.com/dotnet/runtime/issues/37837 tracks the remaining work.

sbomer commented:
Could someone check that this change makes sense? https://github.com/dotnet/runtime/pull/37780/files#diff-20ad030339df429b16201fc3fea8c084R221

sbomer commented on src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs:
```diff
@@ -217,12 +218,12 @@ public sealed class AssemblyBuilder : Assembly
         private bool manifest_module_used;
 
         [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        [DynamicDependency("RuntimeResolve", typeof(ModuleBuilder))]
```
I've moved this attribute closer to the internalcall that depends on it. Does that make sense?

eerhardt commented on src/libraries/System.Net.Primitives/src/System/Net/Cookie.cs:
```diff
@@ -80,24 +79,24 @@ static Cookie()
         }
 #endif
 
-        // These PreserveDependency attributes are a workaround for https://github.com/dotnet/runtime/issues/19348.
```
Not for you to fix here, but reading this message, it sounds like we should move these to a “library build” xml descriptor file instead. That way we don’t need to hack around this with ILKeepDepsAttributes=false. I’ll open an issue for this on Monday.

eerhardt commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactory.cs:
```diff
@@ -25,23 +25,23 @@ public override bool CanConvert(Type typeToConvert)
-        [PreserveDependency(".ctor", "System.Text.Json.Serialization.Converters.ListOfTConverter`2")]
-        [PreserveDependency(".ctor", "System.Text.Json.Serialization.Converters.QueueOfTConverter`2")]
-        [PreserveDependency(".ctor", "System.Text.Json.Serialization.Converters.StackOfTConverter`2")]
+        [DynamicDependency("#ctor", typeof(ArrayConverter<,>))]
```
These should be converted to DynamicallyAccessedMembers attribute. 

FYI @layomia

eerhardt reviewed (approved):
LGTM

sbomer committed:
Include internal attribute for OOB packages

On down-level platforms where DynamicDependencyAttribute doesn't exist
in corelib, this includes it in the OOB assemblies.

sbomer commented:
The remaining [failures](https://helix.dot.net/api/2019-06-17/jobs/b4d42f9d-0b2b-4d48-94b7-a586431db682/workitems/JIT.HardwareIntrinsics/console) look like https://github.com/dotnet/runtime/issues/37946.
